### PR TITLE
Add wte factorization

### PR DESCRIPTION
**Pull request by klei22**
This is a draft for experimenting with new variations on embedding tables.

### CODE REVIEW

**Comment by klei22** [comment]
1. Dataset preparation:
- `cd data/aozorabunko_clean`
- `bash get_dataset.sh` - obtains dataset (requires `jq` command)
- `bash process.sh` - takes a while
- `python3 prepare.py -t input.txt --method char`

2. Manual embedding table creation:
- `python3 mapping.py` maps from the table to an .npy file. 

3. `train.py` and `model.py` setup: 
- `model.py` uses this file to set the word table and lm head if set up in the GPT class constructor
- `train.py`'s `--n_embd_main` needs to match the embedding dimension of the npy matrix

File: gpt_conf.py
Comment (gkielian) [inline gpt_conf.py:1]: Let's discuss the factored embedding table names.

We want to keep n_embd default functionality

So we can have this be the decoder block n_embd...

..let's have the n_embd_wte be None by default
If not none, then let's assume we'll have an upscaling and downscaling adapter.

Depending on the ease of coding, can try initializing randomly first, and later overwrite if an initialization file is specified.

File: mapping.py
Comment (gkielian) [inline mapping.py:1]: This looks more developed, keep this one : )

File: model.py [hunk 1]
```diff
@@ -329,50 +330,76 @@ def forward(self, x):
 
 class GPT(nn.Module):
 
-    def __init__(self, config):
+    def __init__(self, config, wte_path="initial_wte.npy"):
```
Comment (gkielian) [inline model.py:333]: Have wte_path default to None

File: model.py [hunk 2]
```diff
@@ -329,50 +330,76 @@ def forward(self, x):
...
-            ln_f = self.norm_variant_output,
-        ))
+        # Load pre-trained embeddings if a path is provided
+        initial_embeddings = np.load(wte_path)
+        initial_embeddings_tensor = torch.from_numpy(initial_embeddings).float()
+        if wte_path:
+            print("loading wte")
+
+            self.transformer = nn.ModuleDict(dict(
+                wte = nn.Embedding.from_pretrained(initial_embeddings_tensor, freeze=False),
```
Comment (gkielian) [inline model.py:357]: have a config param and train.py for freeze

Comment (gkielian) [reply model.py:357]: preferably we'd be able to freeze select rows, columns or indices, like a freeze mask

File: model.py [hunk 3]
```diff
@@ -329,50 +330,76 @@ def forward(self, x):
...
         if self.softmax_variant_output != "softmax":
             self.softmax_layer_output = softmax_dictionary[config.softmax_variant_output](config)
 
-        self.lm_head = nn.Linear(config.n_embd, config.vocab_size, bias=False)
-        # with weight tying when using torch.compile() some warnings get generated:
-        # "UserWarning: functional_call was passed multiple values for tied weights.
-        # This behavior is deprecated and will be an error in future versions"
-        # not 100% sure what this is, so far seems to be harmless. TODO investigate
-        self.transformer.wte.weight = self.lm_head.weight # https://paperswithcode.com/method/weight-tying
+        self.lm_head = nn.Linear(self.config.n_embd_main, config.vocab_size, bias=False)
```
Comment (gkielian) [inline model.py:383]: Since this is a Linear, we should guard against initialization as well.

File: model.py [hunk 4]
```diff
@@ -329,50 +330,76 @@ def forward(self, x):
...
 
-        # init all weights
+        # Initialize all weights
         self.apply(self._init_weights)
-        # apply special scaled init to the residual projections, per GPT-2 paper
+
+        with torch.no_grad():
+            self.lm_head.weight.copy_(initial_embeddings_tensor)
+        self.transformer.wte.weight = self.lm_head.weight  # https://paperswithcode.com/method/weight-tying
```
Comment (gkielian) [inline model.py:390]: This is where we should set the manual embedding table if specified., right before the setting of the lm_head

File: model.py [hunk 5]
```diff
@@ -329,50 +330,76 @@ def forward(self, x):
...
+
+        print(self.lm_head.weight)
+        print(self.transformer.wte.weight)
+        # Freeze lm_head and wte
+        # self.freeze_layers(['lm_head', 'wte'])
+        # Apply special scaled init to the residual projections, per GPT-2 paper
         for pn, p in self.named_parameters():
             if pn.endswith('c_proj.weight'):
```
Comment (gkielian) [inline model.py:398]: Not part of this commit, but might want to update nomenclature if this isn't reflected in the Block

File: model.py [hunk 6]
```diff
@@ -329,50 +330,76 @@ def forward(self, x):
...
-                torch.nn.init.normal_(p, mean=0.0, std=0.02/math.sqrt(2 * config.n_layer))
+                torch.nn.init.normal_(p, mean=0.0, std=0.02 / math.sqrt(2 * config.n_layer))
 
-        # report number of parameters
-        print("number of parameters: %.2fM" % (self.get_num_params()/1e6,))
+        # Report the number of parameters
+        print("number of parameters: %.2fM" % (self.get_num_params() / 1e6,))
 
     def get_num_params(self, non_embedding=True):
```
Comment (gkielian) [inline model.py:404]: Aside as well, but we should separately report the embedding too

File: train.py [hunk 7]
```diff
@@ -56,6 +59,7 @@ def parse_args():
     model_group.add_argument('--n_head', default=6, type=int)
     model_group.add_argument('--n_kv_group', default=6, type=int)
     model_group.add_argument('--n_embd', default=384, type=int)
+    model_group.add_argument('--n_embd_main', default=24, type=int)
```
Comment (gkielian) [inline train.py:62]: n_embd_wte

File: train.py [hunk 8]
```diff
@@ -30,6 +30,9 @@ def parse_args():
     training_group = parser.add_argument_group('training_group')
     logging_group = parser.add_argument_group('logging_group')
 
+    # Export args
+    training_group.add_argument('--export_embedding', default=None, type=str, help='Path to export the embedding table as a .npy file')
```
Comment (gkielian) [inline train.py:34]: Let's export by default to the same directory as the out file -- but would be more of a run_experiments.py setting

**Review by gkielian — CHANGES_REQUESTED** [review]
exciting, looking forward to merging!
